Libraries/MiscDrivers/SRAM/N01S830HA.c

```diff
@@ -0,0 +1,177 @@
+
+    ERR_CHECK(N01S830HA_write_mode_reg(0b1)); // Disable hold function
+
+    // Now, validate that we were able to write to the mode register
+    // This is the closest thing we have to a "read id" or
+    // communication verification for this SRAM chip.
+    uint8_t mode_reg;
+    ERR_CHECK(N01S830HA_read_mode_reg(&mode_reg));
+    if (mode_reg != 0x01) {
+        return E_NO_DEVICE;
+    }
+
+    // Set burst mode (mode reg bits [7:6])
+    // [7:6] = 0b00 -> word mode
+    // [7:6] = 0b10 -> page mode
+    // [7:6] = 0b01 -> burst mode (*)
+    // [7:6] = 0b11 - > reserved
+
+    // Burst mode allows writing across the 8-bit word and
+    // 32-word page boundaries by automatically incrementing and
+    // wrapping the address pointer.
+    mode_reg &= ~(0b11 << 6);
+    mode_reg |= (0b01 << 6);
+    ERR_CHECK(N01S830HA_write_mode_reg(mode_reg));
+
+    // Validate the mode reg is updated successfully.
+    uint8_t validate = 0;
+    ERR_CHECK(N01S830HA_read_mode_reg(&validate));
+    if (validate != mode_reg) {
+        printf("Expected %i, received %i\n", mode_reg, validate);
+        return E_COMM_ERR;
+    }
+
+    // Re-configure SPI pins now that the hold function has been disabled.
+    // This allows us to use QSPI.
+    ERR_CHECK(spi_init());
+
+    return E_NO_ERROR;
+}
+
+int N01S830HA_enter_quadmode()
+{
+    uint8_t tx_data = CMD_ENABLE_QUAD_IO;
+
+    ERR_CHECK(spi_exit_quadmode());
+    ERR_CHECK(spi_transmit(&tx_data, 1, NULL, 0, true));
+    ERR_CHECK(spi_enter_quadmode());
+
+    g_current_mode = QUAD_MODE;
+
+    return E_NO_ERROR;
+}
+
+int N01S830HA_exit_quadmode()
+{
+    uint8_t tx_data = CMD_RESET_IO;
+
+    ERR_CHECK(spi_enter_quadmode());
+    ERR_CHECK(spi_transmit(&tx_data, 1, NULL, 0, true));
+    ERR_CHECK(spi_exit_quadmode());
+
+    g_current_mode = STANDARD_MODE;
+
+    return E_NO_ERROR;
+}
+
+int N01S830HA_read(uint32_t address, uint8_t *out, unsigned int len)
+{
+    if (g_current_mode == STANDARD_MODE) {
+        _transmit_spi_header(CMD_READ, address);
+    } else if (g_current_mode == QUAD_MODE) {
+        // QUAD mode requires 1 dummy byte.
+        uint8_t header[5]; // (1 byte cmd + 3 byte address + 1 dummy)
+        memset(header, 0xFF, 5);
+        _parse_spi_header(CMD_READ, address, header);
+        ERR_CHECK(spi_transmit(header, 5, NULL, 0, false));
+    }
+
+    return spi_transmit(NULL, 0, out, len, true);
+}
+
+int N01S830HA_write(uint32_t address, uint8_t *data, unsigned int len)
+{
+    _transmit_spi_header(CMD_WRITE, address);
+    return spi_transmit(data, len, NULL, 0, true);
+}
+
+int N01S830HA_write_mode_reg(uint8_t val)
+{
+    uint8_t data[2] = { CMD_WRITE_MODE_REG, val };
+    ERR_CHECK(spi_transmit(data, 2, NULL, 0, true));
+    MXC_Delay(MXC_DELAY_USEC(
+        100)); // Some small delay after updating the mode reg appears to be necessary.
+    return E_NO_ERROR;
+}
+
+int N01S830HA_read_mode_reg(uint8_t *out)
```

Comment:
Just a thought, is there any use for a general read/write_reg function?

Reply:
For this IC no, the mode reg is the only register on the chip